plutus-contract-exe/src/Language/Plutus/Contract/Transaction.hs

```diff
@@ -0,0 +1,171 @@
+module Language.Plutus.Contract.Transaction(
+      UnbalancedTx
+    , emptyTx
+    , computeBalance
+    , inputs
+    , outputs
+    , forge
+    , requiredSignatures
+    , validityRange
+    , mergeWith
+    , toLedgerTx
+    -- * Constructing transactions
+    , unbalancedTx
+    , payToScript
+    , collectFromScript
+    , collectFromScriptFilter
+    ) where
+
+import           Control.Lens      (at, view, (^.))
+import qualified Control.Lens.TH   as Lens.TH
+import qualified Data.Aeson        as Aeson
+import qualified Data.Map          as Map
+import           Data.Maybe        (fromMaybe)
+import qualified Data.Set          as Set
+import           GHC.Generics      (Generic)
+
+import           Ledger            (Address, DataScript, PubKey, RedeemerScript, TxOut, TxOutRef, ValidatorScript)
+import qualified Ledger            as L
+import           Ledger.AddressMap (AddressMap)
+import qualified Ledger.Interval   as I
+import           Ledger.Slot       (SlotRange)
+import qualified Ledger.Tx         as Tx
+import           Ledger.Value      as V
+
+-- | An unsigned and potentially unbalanced transaction, as produced by
+--   a contract endpoint. See note [Unbalanced transactions]
+data UnbalancedTx = UnbalancedTx
+        { _Inputs             :: [(L.TxIn, V.Value)]
+        , _Outputs            :: [L.TxOut]
+        , _Forge              :: V.Value
+        , _RequiredSignatures :: [PubKey]
+        , _ValidityRange      :: SlotRange
+        }
+        deriving stock (Eq, Show, Generic)
+        deriving anyclass (Aeson.FromJSON, Aeson.ToJSON)
+
+Lens.TH.makeLenses ''UnbalancedTx
+
+-- | An unbalanced transaction wiht no inputs and outputs, and an unbounded
+--   validity range.
+--
+emptyTx :: UnbalancedTx
+emptyTx = UnbalancedTx [] [] mempty [] I.always
+
+-- | The ledger transaction of the 'UnbalancedTx'. Note that the result
+--   does not have any signatures, and is potentially unbalanced (ie. invalid).
+--   To produce a balanced 'Tx', use 'Language.Plutus.Contract.Wallet.balance'.
+toLedgerTx :: UnbalancedTx -> L.Tx
+toLedgerTx utx = L.Tx
+            { L.txInputs = Set.fromList (fst <$> _Inputs utx)
+            , L.txOutputs = _Outputs utx
+            , L.txForge = _Forge utx
+            , L.txFee = 0
+            , L.txValidRange = _ValidityRange utx
+            , L.txSignatures = Map.empty
+            }
+
+-- | Compute the difference between the value of the inputs consumed and the
+--   value of the outputs produced by the transaction. If the result is zero
+--   then the transaction is balanced.
+computeBalance :: UnbalancedTx -> V.Value
+computeBalance utx = right `V.minus` left where
+    left = (utx ^. forge) `V.plus` foldMap snd (utx ^. inputs)
+    right = foldMap (view Tx.outValue) (utx ^. outputs)
+
+-- | Combine two unbalanced transactions by appending their respective inputs,
+--   outputs, and signatures, adding their forged values, and applying the
+--   given function to their two validity ranges.
```

Comment:
Seems to me that taking the intersection of the validity ranges is the only sensible thing to do. If we do that then this is sensibly a transaction that can stand in for both the provided transactions.